### DISCUSSION TITLE
Nesting same views in hierarchy best practice

### DISCUSSION THREAD

**Discussion by kakhaberikiknadze** [pointfreeco/swift-composable-architecture#2020] [Q&A] [answered]
What would be the best practice to nest same kind of views together? Let's say a folder app where you've got the list of folders, you tap a folder and it pushes another folder view which is the same (View A -> View A - View A - ViewB for example).

If I'm not missing something reducer will need to have an optional state of the other reducer but since it's the same type of reducer  there will be a recursion error.

``` Swift
Value type 'MyReducer.State' cannot have a stored property that recursively contains it
```

``` Swift
struct MyReducer: ReducerProtocol {
  struct State: Equatable {
     var nested: MyReducer.State? // Error error error 
     var showsNested: Bool {  nested != nil } 
     // .. 
  }

  enum Action {
    case showNested(Bool)
  }
}
```

**Comment by mbrandonw** [comment, marked as answer]
Hi @kakhaberikiknadze, in order for a struct to hold onto an optional value of itself you need to move the value off the stack and onto the heap. You can do this by making a kind of `Box` wrapper type, but also our new `navigation-beta` tools bake this in for free using the `@PresentationState` property wrapper.

**Reply by kakhaberikiknadze** [reply]
That's interesting. I'll have a look 👍🏻 Thanks!

**Reply by kakhaberikiknadze** [reply]
Could you link the repo here? I couldn't find `navigation-beta`. Or are you referring to `swiftui-navigation`?

**Reply by mbrandonw** [reply]
Oh sorry, it's not a repo. It's a branch. Check out the announcement for the navigation beta [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/1944).

**Reply by kakhaberikiknadze** [reply]
Oh, gotcha 🙌🏻 Will have a look now 🤩

**Reply by kakhaberikiknadze** [reply]
Haven't checked the details yet but that looks pretty awesome 🤘🏻 Those overloads, no more need for manually dismissing the child state and such ❤️ Sweet!

**Reply by kakhaberikiknadze** [reply]
Couldn't wait for tomorrow and tested it out. Worked like charm 🥇 Removed some boilerplate as well.

This is how my state and action looks like now

``` Swift
struct MyReducer: RedcerProtocol {
    struct State: Equatable, Then {
        var projects: ProjectsReducer.State
        @PresentationState var directory: ProjectsCoordinatorReducer.State?
        var title: String
        var icon: ImageSource
        @BindingState var videoFile: File?
    }
    
    indirect enum Action: BindableAction {
        case binding(BindingAction<State>)
        case projects(ProjectsReducer.Action)
        case openDirectory(PresentationAction<ProjectsCoordinatorReducer.Action>)
    }
```
